### PR TITLE
Add RenameSymbol command

### LINKED ISSUE #261: Add RenameSymbol command
A new command that prompts the user for a new name and replaces all occurrences of the symbol in the project.

Example from documentation in emacs https://github.com/clojure-emacs/clj-refactor.el/wiki/cljr-rename-symbol

### PR DESCRIPTION
**Pull request by nbardiuk**
Closes #261

### CODE REVIEW

File: test/nrepl_refactor.vim [hunk 1]
```diff
@@ -319,4 +319,64 @@ function! s:suite.add_arity_defmethod_test() abort
...
+        \{'occurrence': '{:file "'.refer_file.'" :line-beg 1 :col-beg 31}'},
+        \{'occurrence': '{:file "'.refer_file.'" :line-beg 3 :col-beg 12}'},
+        \{'status': ['done']}]
+  call s:ch.mock({'status_value': 'open', 'relay':
+        \{m -> get(responses, m['op'], {'status': ['done']})}})
+
+
+  call s:io.mock({'input': 'new-name'})
+  call iced#nrepl#refactor#rename_symbol('dummy')
+  sleep 1000m " TODO figure out how to wait for finished refactoring
```
Comment (nbardiuk) [inline test/nrepl_refactor.vim:361]: @liquidz could you please help me to figure out how to run the test without sleep?
It looks like without the sleep assertions runs before refactoring have chance to write a file.

Comment (liquidz) [reply test/nrepl_refactor.vim:361]: When `iced#nrepl#refactor#rename_symbol` returns a Promise object, you can wait finishing refactoring as below.
```vim
let p = iced#nrepl#refactor#rename_symbol('dummy')
call iced#promise#wait(p)
```

**Review by liquidz — COMMENTED** [review]
@nbardiuk Thanks for your contribution!
I added only comments about `sleep` in test code.

**Comment by nbardiuk** [comment]
@liquidz could you please review the PR?

The most complex technical part for me is to perform edits because `find-symbol` reports the position of the whole defining form instead of position symbol name.

Also I didn't pay attention to usability yet. There are a lot of small interactions like prompt to install jet or reporting/navigation the user after renaming is done. I am open to any suggestion.

**Review by liquidz — CHANGES_REQUESTED** [review]
@nbardiuk Awesome work!
Thank you so much also updating documents! It’s so helpful!

Overall the PR looks good.
I commented on a only few minor points.

In addition, Could you add a `RenameSymbol` to the command palette, and update CHANGELOG.adoc?
* https://github.com/liquidz/vim-iced/blob/2.2.0/autoload/iced/palette.vim
* https://github.com/liquidz/vim-iced/blob/2.2.0/CHANGELOG.adoc

File: autoload/iced/nrepl/refactor.vim [hunk 2]
```diff
@@ -375,5 +376,92 @@ function! iced#nrepl#refactor#add_arity() abort
...
+  endif
+
+  let ns = a:var['ns']
+  let name = a:var['name']
+  let file = a:var['file']
+  let column = a:var['column']
+  let line = a:var['line']
+
+  " find_symbol prints exception for file a jar
+  if s:S.starts_with(file, 'zipfile:/')
```
[the remark this answers is not in the document]
Comment (nbardiuk) [reply autoload/iced/nrepl/refactor.vim]: replaced with `stridx`

File: autoload/iced/nrepl/refactor.vim [hunk 3]
```diff
@@ -375,5 +376,92 @@ function! iced#nrepl#refactor#add_arity() abort
   endtry
 endfunction " }}}
 
+" iced#nrepl#refactor#rename_symbol {{{
+function! iced#nrepl#refactor#rename_symbol(symbol) abort
+  return iced#promise#call('iced#nrepl#var#get', [a:symbol])
```
[the remark this answers is not in the document]
Comment (nbardiuk) [reply autoload/iced/nrepl/refactor.vim:381]: added the `fetching` message, it is really needed on a bigger project

File: test/nrepl_refactor.vim [hunk 4]
```diff
@@ -319,4 +319,154 @@ function! s:suite.add_arity_defmethod_test() abort
...
+        \'file': 'file',
+        \'line': 1,
+        \'column': 1,
+        \'status': ['done']}
+endfunction
+
+function! s:nrepl_mock(ops) abort
+call s:ch.mock({
+      \'status_value': 'open',
+      \'relay': {msg -> get(a:ops, msg['op'], {'status': ['done']})}})
```
[the remark this answers is not in the document]
Comment (nbardiuk) [reply test/nrepl_refactor.vim]: fixed

File: test/nrepl_refactor.vim [hunk 5]
```diff
@@ -319,4 +319,154 @@ function! s:suite.add_arity_defmethod_test() abort
...
+  call s:assert.equals(
+        \ s:io.get_last_args()['echomsg']['text'],
+        \ iced#message#get('not_found'),
+        \ )
+endfunction
+
+function! s:suite.rename_symbol_in_jar_test() abort
+  let info = s:dummy_info('bar')
+  let info.file = 'zipfile:/path/to.jar::some/namespace.clj'
+  call s:nrepl_mock({'info': info})
```
[the remark this answers is not in the document]
Comment (nbardiuk) [reply test/nrepl_refactor.vim:463]: added

File: test/nrepl_refactor.vim [hunk 6]
```diff
@@ -319,4 +319,154 @@ function! s:suite.add_arity_defmethod_test() abort
...
+  call delete(def_file)
+
+  call s:assert.equals(
+        \ s:io.get_last_args()['echomsg']['text'],
+        \ iced#message#get('canceled'),
+        \ )
+endfunction
+
+function! s:suite.rename_symbol_not_found_test() abort
+  call s:nrepl_mock({'info': {'status': ['done', 'no-info']}})
```
[the remark this answers is not in the document]
Comment (nbardiuk) [reply test/nrepl_refactor.vim:449]: added

File: autoload/iced/nrepl/refactor.vim [hunk 7]
```diff
@@ -375,5 +376,92 @@ function! iced#nrepl#refactor#add_arity() abort
...
+    return iced#message#info('canceled')
+  endif
+
+  let symbols = filter(a:symbols, {i, v -> has_key(v, 'occurrence')})
+  let occurrences = map(symbols, {i, v -> iced#promise#sync(edn.decode, [v['occurrence']])})
+
+  " occurrence to the right should be renamed first to avoid shifting column numbers
+  call sort(occurrences, {a, b -> b['col-beg'] - a['col-beg']}) 
+
+  call map(occurrences, {i, v -> s:rename_occurrence(a:old_name, new_name, v)})
```
[the remark this answers is not in the document]
Comment (nbardiuk) [reply autoload/iced/nrepl/refactor.vim]: agree

**Comment by nbardiuk** [comment]
After applying suggestions I've noticed a couple issues:
 * Applying RenameSymbol from Palete hides `Fetching...` message. Don't understand why it happens and whether it should be addressed in scope of this PR
 * `iced#util#restore_context` does not have effect on my neovim config. I will try to figure out if it is my configuration or something wrong with this PR

**Review by liquidz — APPROVED** [review]
@nbardiuk Thanks!
I merge to `dev` branch for now, but I'll cut a new release soon!

**Pull request merged**